feat(linter): implement config for all tsgolint rules supporting options

camchenry opened the pull request:
NOTE: This was largely generated by AI. I've looked over its work and it seems mostly correct, but please do apply some additional scrutiny. This is implementing the boilerplate to add all of the config parsing and docs generation code so we can fully support configuring these rules for tsgolint.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15659?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15659** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15659?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15659?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15286** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15286?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15659).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

connorshea commented on crates/oxc_linter/src/rules/typescript/only_throw_error.rs:
```diff
@@ -1,9 +1,34 @@
+    #[serde(default = "default_true")]
+    pub allow_throwing_any: bool,
+    /// Whether to allow throwing values typed as `unknown`.
+    #[serde(default = "default_true")]
```
I can't remember, are these default_trues necessary? I thought the docs were able to determine this via the Default impl automatically.

camchenry replied:
It seems like the docs can generate just fine without it. Although I think it's probably good to include the information on the struct itself. In this case, `Serialize` requires `Default` anyway, so 🤷

camchenry commented on crates/oxc_linter/src/rules/typescript/no_unnecessary_boolean_literal_compare.rs:
```diff
@@ -1,9 +1,40 @@
+    pub allow_comparing_nullable_booleans_to_true: bool,
+    /// Whether to allow this rule to run without `strictNullChecks` enabled.
+    /// This is not recommended as the rule may produce incorrect results.
+    pub allow_rule_to_run_without_strict_null_checks_i_know_what_i_am_doing: bool,
```
I'm going to remove it from this PR. I'm considering if we should commit to just removing it from `tsgolint`, considering it'll be gone in `typescript-eslint` soon anyway: https://github.com/oxc-project/tsgolint/issues/425

camchenry commented:
### Merge activity

* **Nov 15, 1:54 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Nov 15, 1:55 AM UTC**: `camchenry` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

camchenry committed:
feat(linter): implement config for all tsgolint rules supporting options (#15659)

NOTE: This was largely generated by AI. I've looked over its work and it seems mostly correct, but please do apply some additional scrutiny. This is implementing the boilerplate to add all of the config parsing and docs generation code so we can fully support configuring these rules for tsgolint.
